src/ui/ui_i18n/README.md

```diff
@@ -0,0 +1,169 @@
+and works for all CLDR languages which have pluralization rules
+defined. `intl-messageformat` package exposes `IntlMessageFormat`
+constructor. Messages are provided into the constructor as a
+string message, or a pre-parsed AST object.
+
+```
+import IntlMessageFormat from 'intl-messageformat';
+
+const msg = new IntlMessageFormat(message, locales, [formats]);
+```
+
+The string `message` is parsed, then stored internally in a
+compiled form that is optimized for the `format()` method to
+produce the formatted string for displaying to the user.
+
+```
+const output = msg.format(values);
+```
+
+Creating instances of `IntlMessageFormat` is an expensive operation.
+[Intl-format-cache](https://github.com/yahoo/intl-format-cache)
+library is simply to make it easier to create a cache of format
+instances of a particular type to aid in their reuse. Under the
+hood, this package creates a cache key based on the arguments passed
+to the memoized constructor.
+
+```
+import memoizeIntlConstructor from 'intl-format-cache';
+
+const getMessageFormat = memoizeIntlConstructor(IntlMessageFormat);
+```
+
+`Intl-messageformat` package assumes that the
+[Intl](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl)
+global object exists in the runtime. `Intl` is present in all modern
+browsers and Node.js > 0.10. In order to load `Intl MessageFormat`
+in Node.js we should simply `require()` this package (in Node.js,
+the data for all 200+ languages is loaded along with the library):
+```
+const IntlMessageFormat = require('intl-messageformat');
+```
+
+After that we are able to use this library in the same way as for browsers.
+
+## Angular wrapper
+
+Angular wrapper will have at least 3 entities: translation `service`,
+`directive` and `filter`. Both, the directive and the filter will use
+the translation service with `intl-messageformat` library under the hood.
+
+The translation `service` will provide the following methods (names can be changed):
+- `addMessages(messages: object, [locale: string])` - provides a way to register
+translations with the library
+- `getMessages()` - returns messages for the current language
+- `locale(locale: string)` - tells the library which language to use by given
+language key
+- `locale()` - returns the current locale 
+- `defaultLocale(locale: string)` - tells the library which language to fallback
+when missing translations
+- `defaultLocale()` - returns the default locale
+- `defineFormats(formats: object)` - supply a set of options to the underlying formatter
+- `translate({id: string, values: object, defaultMessage: string})` – translate message by id
+
+The translation `filter` will look like this:
+```
+{{'translationId' | i18n[:{ values: object, defaultMessage: string }]}}
+```
+
+Where:
+- `translationId` - translation id to be translated
+- `values` - values to pass into translation
+- `defaultMessage` - will be used unless translation was successful
```

Comment:
__question:__ `defaultMessage` will be used only if string isn't presented in both target (zh-CN, ru-RU, etc.) __and__ fallback (en) locales, right?

Reply:
> There shouldn't be en messages defined in some translations file, right?

Sorry, I'm not following, do you mean we don't need `en.json` (language file with English translations) if we have default/en messages in the code itself? 

I can think of at least two reasons why we need `en.json`:

* We want to add comments for some labels to give context to the translators (e.g. some labels shouldn't be translated, some are translated differently depending on the context)
* Translators should have access to all labels at once (we can potentially generate them on demand and then apply changes back, but that feels weird)

I was always thinking about `defaultMessage` in the code just as a way to make developers' life easier so that they can see what the actual label is without reading of `en.json`....

Reply:
I think we should try to avoid a separate en.json file entirely, if possible. In practice, the defaultMessage is always the same as the real english message, otherwise either our translation or the defaultMessage is outdated, and either case is not a good situation for us to be in.

We want tools to be able to see new/removed translations, right? That same automation should be able to generate an entirely empty translation file for a new language, in theory. Heck, we could even prefill it with the defaultMessage values. But copying+pasting or referring to some hardcoded en.json file doesn't need to be the way we translators work.

Comments are a little harder, but since the comments are about the context in which the translation is used, perhaps that's something we should be baking into the components themselves, so a person can maintain an optional "context comment" alongside the default message and whatnot.

Reply:
@azasypkin, maybe I'm not understanding the question or context here, but an en.json is always the baseline that localization works off of. We would start with an en.json to send out the content for first-time translation, and for updates, the diff file between en.json v1 to en.json v2 is what will be used to determine what to update in the other languages.

Reply:
@epixa ah, got it. In that case, no, I don't think we need a static en.json file in the repository. We wouldn't have json files in the other language repos either, right?

Reply:
We would have static copies of the other languages checked into the repo. The reason we don't need to do that for en.json is that one of our technical requirements is to have defaultMessage in the templates themselves, and that message will always be english. An en.json file would just be all of those exact same messages aggregated into one place, which we can do through automation rather than hardcoding.

Reply:
@epixa, it's helpful to have an equivalent English repo because it's faster to compare between EN vs other language, to find the source English when there are questions from linguists. Then I (or whoever is checking files) could easily compare without needing a dev to run a JSON export. Or it's easier to identify when for example, there are discrepancies in files due to version control (sometimes not everything in English is sent for localizing, but may be added or removed at a different date), or sometimes from bugs in scripts used to process the files, etc.

Reply:
I can see how duplicating messages can cause problems. Having matching repos is definitely easier to compare between languages and versions, but as long as we have unique keys to identify and locate equivalent messages/strings in English, it doesn't necessarily need to be an en.json file.

Reply:
> then we just have a new pain point with having to continuously identify and reconcile discrepancies between those two things. But this seems like something we should be able to solve through automation.

Yeah, even if we have `en.json`, developers don't have to directly interact with it:  if default message or/and context has been changed in the code then tools will update it (and `json` files for other languages too?). Basically you shouldn't be able to commit if default message in code isn't equal to the string in `en.json`. Also if commit includes updated `en.json` that will probably make it easier to find all PRs between releases that touched localizable text (not a huge deal though if we manually mark every such PR with relevant `i18n-change` labels).

Reply:
@maksim-tolo what are your thoughts on this?

Reply:
++ Agree. It would be nice to have a little explanation about `defaultMessage` being the final fallback, there being a tool that generates `en.json` containing `defaultMessage`s defined inline, and using the changes to `en.json` to trigger manual updates of other languages.

Reply:
Yes, we don't need `en.json` file that duplicates `defaultMessage`s. I've added the explanation about `defaultMessage`s and `en.json` into the doc.